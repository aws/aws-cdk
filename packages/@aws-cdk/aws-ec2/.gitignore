*.js
tsconfig.json
*.js.map
*.d.ts
*.generated.ts
dist
test/cdk.json
.jsii

.LAST_BUILD
.nyc_output
coverage
nyc.config.js
.LAST_PACKAGE
*.snk
!.eslintrc.js
