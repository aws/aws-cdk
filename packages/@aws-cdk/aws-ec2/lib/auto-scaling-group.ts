import autoscaling = require('@aws-cdk/aws-autoscaling');
import iam = require('@aws-cdk/aws-iam');
import sns = require('@aws-cdk/aws-sns');
import cdk = require('@aws-cdk/cdk');
import { AllConnections, AnyIPv4, IConnectionPeer } from './connection';
import { Connections } from './connections';
import { InstanceType } from './instance-types';
import { ClassicLoadBalancer, IClassicLoadBalancerTarget } from './load-balancer';
import { IMachineImageSource, OperatingSystemType } from './machine-image';
import { SecurityGroup } from './security-group';
import { VpcNetworkRef, VpcPlacementStrategy } from './vpc-ref';

/**
 * Properties of a Fleet
 */
export interface AutoScalingGroupProps {
    /**
     * Type of instance to launch
     */
    instanceType: InstanceType;

    /**
     * Minimum number of instances in the fleet
     * @default 1
     */
    minSize?: number;

    /**
     * Maximum number of instances in the fleet
     * @default 1
     */
    maxSize?: number;

    /**
     * Initial amount of instances in the fleet
     * @default 1
     */
    desiredCapacity?: number;

    /**
     * Name of SSH keypair to grant access to instances
     * @default No SSH access will be possible
     */
    keyName?: string;

    /**
     * AMI to launch
     */
    machineImage: IMachineImageSource;

    /**
     * VPC to launch these instances in.
     */
    vpc: VpcNetworkRef;

    /**
     * Where to place instances within the VPC
     */
    vpcPlacement?: VpcPlacementStrategy;

    /**
     * SNS topic to send notifications about fleet changes
     * @default No fleet change notifications will be sent.
     */
    notificationsTopic?: sns.cloudformation.TopicResource;

    /**
     * Whether the instances can initiate connections to anywhere by default
     *
     * @default true
     */
    allowAllOutbound?: boolean;
}

/**
 * A Fleet represents a managed set of EC2 instances
 *
 * The Fleet models a number of AutoScalingGroups, a launch configuration, a
 * security group and an instance role.
 *
 * It allows adding arbitrary commands to the startup scripts of the instances
 * in the fleet.
 *
 * The ASG spans all availability zones.
 */
export class AutoScalingGroup extends cdk.Construct implements IClassicLoadBalancerTarget {
    public readonly connectionPeer: IConnectionPeer;

    /**
     * The type of OS instances of this fleet are running.
     */
    public readonly osType: OperatingSystemType;

    /**
     * Allows specify security group connections for instances of this fleet.
     */
    public readonly connections: Connections;

    /**
     * The IAM role assumed by instances of this fleet.
     */
    public readonly role: iam.Role;

    private readonly userDataLines = new Array<string>();
    private readonly autoScalingGroup: autoscaling.cloudformation.AutoScalingGroupResource;
    private readonly securityGroup: SecurityGroup;
    private readonly loadBalancerNames: cdk.Token[] = [];

    constructor(parent: cdk.Construct, name: string, props: AutoScalingGroupProps) {
        super(parent, name);

        this.securityGroup = new SecurityGroup(this, 'InstanceSecurityGroup', { vpc: props.vpc });
        this.connections = new Connections(this.securityGroup);
        this.connectionPeer = this.securityGroup;

        if (props.allowAllOutbound !== false) {
            this.connections.allowTo(new AnyIPv4(), new AllConnections(), 'Outbound traffic allowed by default');
        }

        this.role = new iam.Role(this, 'InstanceRole', {
            assumedBy: new cdk.ServicePrincipal('ec2.amazonaws.com')
        });

        const iamProfile = new iam.cloudformation.InstanceProfileResource(this, 'InstanceProfile', {
            roles: [ this.role.roleName ]
        });

        // use delayed evaluation
        const machineImage = props.machineImage.getImage(this);
        const userDataToken = new cdk.Token(() => new cdk.FnBase64((machineImage.os.createUserData(this.userDataLines))));

        const launchConfig = new autoscaling.cloudformation.LaunchConfigurationResource(this, 'LaunchConfig', {
            imageId: machineImage.imageId,
            keyName: props.keyName,
            instanceType: props.instanceType.toString(),
            securityGroups: [this.securityGroup.securityGroupId],
            iamInstanceProfile: iamProfile.ref,
            userData: userDataToken
        });

        launchConfig.addDependency(this.role);

        const minSize = props.minSize || 1;
        const maxSize = props.maxSize || 1;
        const desiredCapacity = props.desiredCapacity || 1;

        const asgProps: autoscaling.cloudformation.AutoScalingGroupResourceProps = {
            minSize: minSize.toString(),
            maxSize: maxSize.toString(),
            desiredCapacity: desiredCapacity.toString(),
            launchConfigurationName: launchConfig.ref,
            loadBalancerNames: new cdk.Token(() => this.loadBalancerNames),
        };

        if (props.notificationsTopic) {
            asgProps.notificationConfigurations = [];
            asgProps.notificationConfigurations.push({
                topicArn: props.notificationsTopic.ref,
                notificationTypes: [
                    "autoscaling:EC2_INSTANCE_LAUNCH",
                    "autoscaling:EC2_INSTANCE_LAUNCH_ERROR",
                    "autoscaling:EC2_INSTANCE_TERMINATE",
                    "autoscaling:EC2_INSTANCE_TERMINATE_ERROR"
                ],
            });
        }

        const subnets = props.vpc.subnets(props.vpcPlacement);
        asgProps.vpcZoneIdentifier = subnets.map(n => n.subnetId);

        this.autoScalingGroup = new autoscaling.cloudformation.AutoScalingGroupResource(this, 'ASG', asgProps);
        this.osType = machineImage.os.type;
    }

    public attachToClassicLB(loadBalancer: ClassicLoadBalancer): void {
        this.loadBalancerNames.push(loadBalancer.loadBalancerName);
    }

    /**
     * Add command to the startup script of fleet instances.
     * The command must be in the scripting language supported by the fleet's OS (i.e. Linux/Windows).
     */
    public addUserData(script: string) {
        this.userDataLines.push(script);
    }

    public autoScalingGroupName() {
        return this.autoScalingGroup.ref;
    }

    /**
     * Adds a statement to the IAM role assumed by instances of this fleet.
     */
    public addToRolePolicy(statement: cdk.PolicyStatement) {
        this.role.addToPolicy(statement);
    }
}
