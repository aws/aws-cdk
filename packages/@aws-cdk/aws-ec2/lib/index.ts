export * from './bastion-host';
export * from './connections';
export * from './cfn-init';
export * from './cfn-init-elements';
export * from './instance-types';
export * from './instance';
export * from './launch-template';
export * from './machine-image';
export * from './nat';
export * from './network-acl';
export * from './network-acl-types';
export * from './port';
export * from './security-group';
export * from './subnet';
export * from './peer';
export * from './volume';
export * from './vpc';
export * from './vpc-lookup';
export * from './vpn';
export * from './vpc-endpoint';
export * from './vpc-endpoint-service';
export * from './user-data';
export * from './windows-versions';
export * from './vpc-flow-logs';
export * from './client-vpn-endpoint-types';
export * from './client-vpn-endpoint';
export * from './client-vpn-authorization-rule';
export * from './client-vpn-route';

// AWS::EC2 CloudFormation Resources:
export * from './ec2.generated';

import './ec2-augmentations.generated';
