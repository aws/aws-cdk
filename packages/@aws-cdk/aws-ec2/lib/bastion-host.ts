import { IPrincipal, IRole, PolicyStatement } from '@aws-cdk/aws-iam';
import { CfnOutput, Construct, Stack } from '@aws-cdk/core';
import { AmazonLinuxGeneration, InstanceClass, InstanceSize, InstanceType } from '.';
import { Connections } from './connections';
import { IInstance, Instance } from './instance';
import { IMachineImage, MachineImage } from './machine-image';
import { IPeer } from './peer';
import { Port } from './port';
import { ISecurityGroup } from './security-group';
import { IVpc, SubnetSelection } from './vpc';

/**
 * Properties of the bastion host
 *
 * @experimental
 */
export interface BastionHostLinuxProps {

  /**
   * In which AZ to place the instance within the VPC
   *
   * @default - Random zone.
   */
  readonly availabilityZone?: string;

  /**
   * VPC to launch the instance in.
   */
  readonly vpc: IVpc;

  /**
   * The name of the instance
   *
   * @default 'BastionHost'
   */
  readonly instanceName?: string;

  /**
   * Select the subnets to run the bastion host in.
   * Set this to PUBLIC if you need to connect to this instance via the internet and cannot use SSM.
   * You have to allow port 22 manually by using the connections field
   *
   * @default - private subnets of the supplied VPC
   */
  readonly subnetSelection?: SubnetSelection;

  /**
   * Security Group to assign to this instance
   *
   * @default - create new security group with no inbound and all outbound traffic allowed
   */
  readonly securityGroup?: ISecurityGroup;

  /**
   * Type of instance to launch
   * @default 't3.nano'
   */
  readonly instanceType?: InstanceType;

  /**
   * The machine image to use
   *
   * @default - An Amazon Linux 2 image which is kept up-to-date automatically (the instance
   * may be replaced on every deployment).
   */
  readonly machineImage?: IMachineImage;
}

/**
 * This creates a linux bastion host you can use to connect to other instances or services in your VPC.
 * The recommended way to connect to the bastion host is by using AWS Systems Manager Session Manager.
 *
 * The operating system is Amazon Linux 2 with the latest SSM agent installed
 *
 * You can also configure this bastion host to allow connections via SSH
 *
 * @experimental
 */
export class BastionHostLinux extends Construct implements IInstance {
  public readonly stack: Stack;

  /**
   * Allows specify security group connections for the instance.
   */
  public readonly connections: Connections;

  /**
   * The IAM role assumed by the instance.
   */
  public readonly role: IRole;

  /**
   * The principal to grant permissions to
   */
  public readonly grantPrincipal: IPrincipal;

  /**
   * The underlying instance resource
   */
  public readonly instance: Instance;

  /**
   * @attribute
   */
  public readonly instanceId: string;

  /**
   * @attribute
   */
  public readonly instanceAvailabilityZone: string;

  /**
   * @attribute
   */
  public readonly instancePrivateDnsName: string;
  /**
   * @attribute
   */
  public readonly instancePrivateIp: string;
  /**
   * @attribute
   */
  public readonly instancePublicDnsName: string;
  /**
   * @attribute
   */
  public readonly instancePublicIp: string;

  constructor(scope: Construct, id: string, props: BastionHostLinuxProps) {
    super(scope, id);
    this.stack = Stack.of(scope);
    this.instance = new Instance(this, 'Resource', {
      vpc: props.vpc,
      availabilityZone: props.availabilityZone,
      securityGroup: props.securityGroup,
      instanceName: props.instanceName ?? 'BastionHost',
      instanceType: props.instanceType ?? InstanceType.of(InstanceClass.T3, InstanceSize.NANO),
      machineImage: props.machineImage ?? MachineImage.latestAmazonLinux({ generation: AmazonLinuxGeneration.AMAZON_LINUX_2 }),
      vpcSubnets: props.subnetSelection ?? {},
    });
    this.instance.addToRolePolicy(new PolicyStatement({
      actions: [
        'ssmmessages:*',
        'ssm:UpdateInstanceInformation',
        'ec2messages:*',
      ],
      resources: ['*'],
    }));
    this.instance.addUserData('yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm');

    this.connections = this.instance.connections;
    this.role = this.instance.role;
    this.grantPrincipal = this.instance.role;
    this.instanceId = this.instance.instanceId;
    this.instancePrivateIp = this.instance.instancePrivateIp;
    this.instanceAvailabilityZone = this.instance.instanceAvailabilityZone;
    this.instancePrivateDnsName = this.instance.instancePrivateDnsName;
    this.instancePublicIp = this.instance.instancePublicIp;
    this.instancePublicDnsName = this.instance.instancePublicDnsName;

    new CfnOutput(this, 'BastionHostId', {
      description: 'Instance ID of the bastion host. Use this to connect via SSM Session Manager',
      value: this.instanceId,
    });
  }

  /**
   * Allow SSH access from the given peer or peers
   *
   * Necessary if you want to connect to the instance using ssh. If not
   * called, you should use SSM Session Manager to connect to the instance.
   */
  public allowSshAccessFrom(...peer: IPeer[]): void {
    peer.forEach(p => {
      this.connections.allowFrom(p, Port.tcp(22), 'SSH access');
    });
  }
}