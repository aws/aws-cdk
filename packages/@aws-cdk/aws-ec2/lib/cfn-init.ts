import * as crypto from 'crypto';
import * as iam from '@aws-cdk/aws-iam';
import { Aws, CfnResource, Construct } from '@aws-cdk/core';
import { InitElement } from './cfn-init-elements';
import { OperatingSystemType } from './machine-image';
import { InitBindOptions, InitElementConfig, InitElementType, InitPlatform } from './private/cfn-init-internal';
import { UserData } from './user-data';

/**
 * A CloudFormation-init configuration
 */
export class CloudFormationInit {
  /**
   * Build a new config from a set of Init Elements
   */
  public static fromElements(...elements: InitElement[]): CloudFormationInit {
    return CloudFormationInit.fromConfig(new InitConfig(elements));
  }

  /**
   * Use an existing InitConfig object as the default and only config
   */
  public static fromConfig(config: InitConfig): CloudFormationInit {
    return CloudFormationInit.fromConfigSets({
      configSets: {
        default: ['config'],
      },
      configs: { config },
    });
  }

  /**
   * Build a CloudFormationInit from config sets
   */
  public static fromConfigSets(props: ConfigSetProps): CloudFormationInit {
    return new CloudFormationInit(props.configSets, props.configs);
  }

  private readonly _configSets: Record<string, string[]> = {};
  private readonly _configs: Record<string, InitConfig> = {};

  private constructor(configSets: Record<string, string[]>, configs: Record<string, InitConfig>) {
    Object.assign(this._configSets, configSets);
    Object.assign(this._configs, configs);
  }

  /**
   * Add a config with the given name to this CloudFormationInit object
   */
  public addConfig(configName: string, config: InitConfig) {
    if (this._configs[configName]) {
      throw new Error(`CloudFormationInit already contains a config named '${configName}'`);
    }
    this._configs[configName] = config;
  }

  /**
   * Add a config set with the given name to this CloudFormationInit object
   *
   * The new configset will reference the given configs in the given order.
   */
  public addConfigSet(configSetName: string, configNames: string[] = []) {
    if (this._configSets[configSetName]) {
      throw new Error(`CloudFormationInit already contains a configSet named '${configSetName}'`);
    }

    const unk = configNames.filter(c => !this._configs[c]);
    if (unk.length > 0) {
      throw new Error(`Unknown configs referenced in definition of '${configSetName}': ${unk}`);
    }

    this._configSets[configSetName] = [...configNames];
  }

  /**
   * Attach the CloudFormation Init config to the given resource
   *
   * As an app builder, use `instance.applyCloudFormationInit()` or
   * `autoScalingGroup.applyCloudFormationInit()` to trigger this method.
   *
   * This method does the following:
   *
   * - Renders the `AWS::CloudFormation::Init` object to the given resource's
   *   metadata, potentially adding a `AWS::CloudFormation::Authentication` object
   *   next to it if required.
   * - Updates the instance role policy to be able to call the APIs required for
   *   `cfn-init` and `cfn-signal` to work, and potentially add permissions to download
   *   referenced asset and bucket resources.
   * - Updates the given UserData with commands to execute the `cfn-init` script.
   */
  public attach(attachedResource: CfnResource, attachOptions: AttachInitOptions) {
    if (attachOptions.platform === OperatingSystemType.UNKNOWN) {
      throw new Error('Cannot attach CloudFormationInit to an unknown OS type');
    }

    const CFN_INIT_METADATA_KEY = 'AWS::CloudFormation::Init';

    if (attachedResource.getMetadata(CFN_INIT_METADATA_KEY) !== undefined) {
      throw new Error(`Cannot bind CfnInit: resource '${attachedResource.node.path}' already has '${CFN_INIT_METADATA_KEY}' attached`);
    }

    // Note: This will not reflect mutations made after attaching.
    const bindResult = this.bind(attachedResource.stack, attachOptions);
    attachedResource.addMetadata(CFN_INIT_METADATA_KEY, bindResult.configData);

    // Need to resolve the various tokens from assets in the config,
    // as well as include any asset hashes provided so the fingerprint is accurate.
    const resolvedConfig = attachedResource.stack.resolve(bindResult.configData);
    const fingerprintInput = { config: resolvedConfig, assetHash: bindResult.assetHash };
    const fingerprint = contentHash(JSON.stringify(fingerprintInput)).substr(0, 16);

    attachOptions.instanceRole.addToPolicy(new iam.PolicyStatement({
      actions: ['cloudformation:DescribeStackResource', 'cloudformation:SignalResource'],
      resources: [Aws.STACK_ID],
    }));

    if (bindResult.authData) {
      attachedResource.addMetadata('AWS::CloudFormation::Authentication', bindResult.authData);
    }

    // To identify the resources that have the metadata and where the signal
    // needs to be sent, we need { region, stackName, logicalId }
    const resourceLocator = `--region ${Aws.REGION} --stack ${Aws.STACK_NAME} --resource ${attachedResource.logicalId}`;
    const configSets = (attachOptions.configSets ?? ['default']).join(',');
    const printLog = attachOptions.printLog ?? true;

    if (attachOptions.embedFingerprint ?? true) {
      // It just so happens that the comment char is '#' for both bash and PowerShell
      attachOptions.userData.addCommands(`# fingerprint: ${fingerprint}`);
    }

    if (attachOptions.platform === OperatingSystemType.WINDOWS) {
      const errCode = attachOptions.ignoreFailures ? '0' : '$LASTEXITCODE';
      attachOptions.userData.addCommands(...[
        `cfn-init.exe -v ${resourceLocator} -c ${configSets}`,
        `cfn-signal.exe -e ${errCode} ${resourceLocator}`,
        ...printLog ? ['type C:\\cfn\\log\\cfn-init.log'] : [],
      ]);
    } else {
      const errCode = attachOptions.ignoreFailures ? '0' : '$?';
      attachOptions.userData.addCommands(...[
        // Run a subshell without 'errexit', so we can signal using the exit code of cfn-init
        '(',
        '  set +e',
        `  /opt/aws/bin/cfn-init -v ${resourceLocator} -c ${configSets}`,
        `  /opt/aws/bin/cfn-signal -e ${errCode} ${resourceLocator}`,
        ...printLog ? ['  cat /var/log/cfn-init.log >&2'] : [],
        ')',
      ]);
    }
  }

  private bind(scope: Construct, options: AttachInitOptions): { configData: any, authData: any, assetHash?: any } {
    const nonEmptyConfigs = mapValues(this._configs, c => c.isEmpty() ? undefined : c);

    const configNameToBindResult = mapValues(nonEmptyConfigs, c => c._bind(scope, options));

    return {
      configData: {
        configSets: mapValues(this._configSets, configNames => configNames.filter(name => nonEmptyConfigs[name] !== undefined)),
        ...mapValues(configNameToBindResult, c => c.config),
      },
      authData: Object.values(configNameToBindResult).map(c => c.authentication).reduce(deepMerge, undefined),
      assetHash: combineAssetHashesOrUndefined(Object.values(configNameToBindResult).map(c => c.assetHash)),
    };
  }

}

/**
 * A collection of configuration elements
 */
export class InitConfig {
  private readonly elements = new Array<InitElement>();

  constructor(elements: InitElement[]) {
    this.add(...elements);
  }

  /**
   * Whether this configset has elements or not
   */
  public isEmpty() {
    return this.elements.length === 0;
  }

  /**
   * Add one or more elements to the config
   */
  public add(...elements: InitElement[]) {
    this.elements.push(...elements);
  }

  /**
   * Called when the config is applied to an instance.
   * Creates the CloudFormation representation of the Init config and handles any permissions and assets.
   * @internal
   */
  public _bind(scope: Construct, options: AttachInitOptions): InitElementConfig {
    const bindOptions = {
      instanceRole: options.instanceRole,
      platform: this.initPlatformFromOSType(options.platform),
      scope,
    };

    const packageConfig = this.bindForType(InitElementType.PACKAGE, bindOptions);
    const groupsConfig = this.bindForType(InitElementType.GROUP, bindOptions);
    const usersConfig = this.bindForType(InitElementType.USER, bindOptions);
    const sourcesConfig = this.bindForType(InitElementType.SOURCE, bindOptions);
    const filesConfig = this.bindForType(InitElementType.FILE, bindOptions);
    const commandsConfig = this.bindForType(InitElementType.COMMAND, bindOptions);
    // Must be last!
    const servicesConfig = this.bindForType(InitElementType.SERVICE, bindOptions);

    const allConfig = [packageConfig, groupsConfig, usersConfig, sourcesConfig, filesConfig, commandsConfig, servicesConfig];
    const authentication = allConfig.map(c => c?.authentication).reduce(deepMerge, undefined);
    const assetHash = combineAssetHashesOrUndefined(allConfig.map(c => c?.assetHash));

    return {
      config: {
        packages: packageConfig?.config,
        groups: groupsConfig?.config,
        users: usersConfig?.config,
        sources: sourcesConfig?.config,
        files: filesConfig?.config,
        commands: commandsConfig?.config,
        services: servicesConfig?.config,
      },
      authentication,
      assetHash,
    };
  }

  private bindForType(elementType: InitElementType, renderOptions: Omit<InitBindOptions, 'index'>): InitElementConfig | undefined {
    const elements = this.elements.filter(elem => elem.elementType === elementType);
    if (elements.length === 0) { return undefined; }

    const bindResults = elements.map((e, index) => e._bind({ index, ...renderOptions }));

    return {
      config: bindResults.map(r => r.config).reduce(deepMerge, undefined) ?? {},
      authentication: bindResults.map(r => r.authentication).reduce(deepMerge, undefined),
      assetHash: combineAssetHashesOrUndefined(bindResults.map(r => r.assetHash)),
    };
  }

  private initPlatformFromOSType(osType: OperatingSystemType): InitPlatform {
    switch (osType) {
      case OperatingSystemType.LINUX: {
        return InitPlatform.LINUX;
      }
      case OperatingSystemType.WINDOWS: {
        return InitPlatform.WINDOWS;
      }
      default: {
        throw new Error('Cannot attach CloudFormationInit to an unknown OS type');
      }
    }
  }
}

/**
 * Options for CloudFormationInit.withConfigSets
 */
export interface ConfigSetProps {
  /**
   * The definitions of each config set
   */
  readonly configSets: Record<string, string[]>;

  /**
   * The sets of configs to pick from
   */
  readonly configs: Record<string, InitConfig>;
}

/**
 * Deep-merge objects and arrays
 *
 * Treat arrays as sets, removing duplicates. This is acceptable for rendering
 * cfn-inits, not applicable elsewhere.
 */
function deepMerge(target?: Record<string, any>, src?: Record<string, any>) {
  if (target == null) { return src; }
  if (src == null) { return target; }

  for (const [key, value] of Object.entries(src)) {
    if (Array.isArray(value)) {
      if (target[key] && !Array.isArray(target[key])) {
        throw new Error(`Trying to merge array [${value}] into a non-array '${target[key]}'`);
      }
      target[key] = Array.from(new Set([
        ...target[key] ?? [],
        ...value,
      ]));
      continue;
    }
    if (typeof value === 'object' && value) {
      target[key] = deepMerge(target[key] ?? {}, value);
      continue;
    }
    if (value !== undefined) {
      target[key] = value;
    }
  }

  return target;
}

/**
 * Map a function over values of an object
 *
 * If the mapping function returns undefined, remove the key
 */
function mapValues<A, B>(xs: Record<string, A>, fn: (x: A) => B | undefined): Record<string, B> {
  const ret: Record<string, B> = {};
  for (const [k, v] of Object.entries(xs)) {
    const mapped = fn(v);
    if (mapped !== undefined) {
      ret[k] = mapped;
    }
  }
  return ret;
}

// Combines all input asset hashes into one, or if no hashes are present, returns undefined.
function combineAssetHashesOrUndefined(hashes: (string | undefined)[]): string | undefined {
  const hashArray = hashes.filter((x): x is string => x !== undefined);
  return hashArray.length > 0 ? hashArray.join('') : undefined;
}

function contentHash(content: string) {
  return crypto.createHash('sha256').update(content).digest('hex');
}

/**
 * Options for attaching a CloudFormationInit to a resource
 */
export interface AttachInitOptions {
  /**
   * Instance role of the consuming instance or fleet
   */
  readonly instanceRole: iam.IRole;

  /**
   * OS Platform the init config will be used for
   */
  readonly platform: OperatingSystemType;

  /**
   * UserData to add commands to
   */
  readonly userData: UserData;

  /**
   * ConfigSet to activate
   *
   * @default ['default']
   */
  readonly configSets?: string[];

  /**
   * Whether to embed a hash into the userData
   *
   * If `true` (the default), a hash of the config will be embedded into the
   * UserData, so that if the config changes, the UserData changes and
   * the instance will be replaced.
   *
   * If `false`, no such hash will be embedded, and if the CloudFormation Init
   * config changes nothing will happen to the running instance.
   *
   * @default true
   */
  readonly embedFingerprint?: boolean;

  /**
   * Print the results of running cfn-init to the Instance System Log
   *
   * By default, the output of running cfn-init is written to a log file
   * on the instance. Set this to `true` to print it to the System Log
   * (visible from the EC2 Console), `false` to not print it.
   *
   * (Be aware that the system log is refreshed at certain points in
   * time of the instance life cycle, and successful execution may
   * not always show up).
   *
   * @default true
   */
  readonly printLog?: boolean;

  /**
   * Don't fail the instance creation when cfn-init fails
   *
   * You can use this to prevent CloudFormation from rolling back when
   * instances fail to start up, to help in debugging.
   *
   * @default false
   */
  readonly ignoreFailures?: boolean;
}
