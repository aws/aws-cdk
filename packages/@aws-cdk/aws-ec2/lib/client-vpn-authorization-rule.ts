import { Resource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { IClientVpnEndpoint } from './client-vpn-endpoint-types';
import { CfnClientVpnAuthorizationRule } from './ec2.generated';

/**
 * Options for a ClientVpnAuthorizationRule
 */
export interface ClientVpnAuthorizationRuleOptions {
  /**
   * The IPv4 address range, in CIDR notation, of the network for which access
   * is being authorized.
   */
  readonly cidr: string;

  /**
   * The ID of the group to grant access to, for example, the Active Directory
   * group or identity provider (IdP) group.
   *
   * @default - authorize all groups
   */
  readonly groupId?: string;

  /**
   * A brief description of the authorization rule.
   *
   * @default - no description
   */
  readonly description?: string;
}

/**
 * Properties for a ClientVpnAuthorizationRule
 */
export interface ClientVpnAuthorizationRuleProps extends ClientVpnAuthorizationRuleOptions {
  /**
   * The client VPN endpoint to which to add the rule.
   */
  readonly clientVpnEndoint: IClientVpnEndpoint;
}

/**
 * A client VPN authorization rule
 */
export class ClientVpnAuthorizationRule extends Resource {
  constructor(scope: Construct, id: string, props: ClientVpnAuthorizationRuleProps) {
    super(scope, id);

    new CfnClientVpnAuthorizationRule(this, 'Resource', {
      clientVpnEndpointId: props.clientVpnEndoint.endpointId,
      targetNetworkCidr: props.cidr,
      accessGroupId: props.groupId,
      authorizeAllGroups: !props.groupId,
      description: props.description,
    });
  }
}
