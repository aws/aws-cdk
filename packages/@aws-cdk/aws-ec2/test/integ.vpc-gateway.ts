import * as cdk from '@aws-cdk/core';
import * as ec2 from '../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-cdk-ec2-vpc-gateway');

const vpc = new ec2.Vpc(stack, 'MyVpc', {
  maxAzs: 1,
  subnetConfiguration: [
    {
      subnetType: ec2.SubnetType.PUBLIC,
      name: 'Public',
    },
    {
      subnetType: ec2.SubnetType.ISOLATED,
      name: 'Isolated',
    },
  ],
});

(vpc.isolatedSubnets[0] as ec2.Subnet).addRoute('MyRoute', {
  routerId: vpc.internetGatewayId!,
  routerType: ec2.RouterType.GATEWAY,
  destinationCidrBlock: '8.8.8.8/32',
});

app.synth();
