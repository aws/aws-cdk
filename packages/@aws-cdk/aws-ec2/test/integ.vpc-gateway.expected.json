{
  "Resources": {
    "MyVpcF9F0CA6F": {
      "Type": "AWS::EC2::VPC",
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "aws-cdk-ec2-vpc-gateway/MyVpc"
          }
        ]
      }
    },
    "MyVpcPublicSubnet1SubnetF6608456": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.0.0/17",
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F"
        },
        "AvailabilityZone": "test-region-1a",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public"
          },
          {
            "Key": "Name",
            "Value": "aws-cdk-ec2-vpc-gateway/MyVpc/PublicSubnet1"
          }
        ]
      }
    },
    "MyVpcPublicSubnet1RouteTableC46AB2F4": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "aws-cdk-ec2-vpc-gateway/MyVpc/PublicSubnet1"
          }
        ]
      }
    },
    "MyVpcPublicSubnet1RouteTableAssociation2ECEE1CB": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcPublicSubnet1RouteTableC46AB2F4"
        },
        "SubnetId": {
          "Ref": "MyVpcPublicSubnet1SubnetF6608456"
        }
      }
    },
    "MyVpcPublicSubnet1DefaultRoute95FDF9EB": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcPublicSubnet1RouteTableC46AB2F4"
        },
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "MyVpcIGW5C4A4F63"
        }
      },
      "DependsOn": [
        "MyVpcVPCGW488ACE0D"
      ]
    },
    "MyVpcIsolatedSubnet1Subnet2259FE9F": {
      "Type": "AWS::EC2::Subnet",
      "Properties": {
        "CidrBlock": "10.0.128.0/17",
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F"
        },
        "AvailabilityZone": "test-region-1a",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated"
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated"
          },
          {
            "Key": "Name",
            "Value": "aws-cdk-ec2-vpc-gateway/MyVpc/IsolatedSubnet1"
          }
        ]
      }
    },
    "MyVpcIsolatedSubnet1RouteTable67AEA7B8": {
      "Type": "AWS::EC2::RouteTable",
      "Properties": {
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F"
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "aws-cdk-ec2-vpc-gateway/MyVpc/IsolatedSubnet1"
          }
        ]
      }
    },
    "MyVpcIsolatedSubnet1RouteTableAssociationCDAE5449": {
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcIsolatedSubnet1RouteTable67AEA7B8"
        },
        "SubnetId": {
          "Ref": "MyVpcIsolatedSubnet1Subnet2259FE9F"
        }
      }
    },
    "MyVpcIsolatedSubnet1MyRouteCDD7D172": {
      "Type": "AWS::EC2::Route",
      "Properties": {
        "RouteTableId": {
          "Ref": "MyVpcIsolatedSubnet1RouteTable67AEA7B8"
        },
        "DestinationCidrBlock": "8.8.8.8/32",
        "GatewayId": {
          "Ref": "MyVpcIGW5C4A4F63"
        }
      }
    },
    "MyVpcIGW5C4A4F63": {
      "Type": "AWS::EC2::InternetGateway",
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "aws-cdk-ec2-vpc-gateway/MyVpc"
          }
        ]
      }
    },
    "MyVpcVPCGW488ACE0D": {
      "Type": "AWS::EC2::VPCGatewayAttachment",
      "Properties": {
        "VpcId": {
          "Ref": "MyVpcF9F0CA6F"
        },
        "InternetGatewayId": {
          "Ref": "MyVpcIGW5C4A4F63"
        }
      }
    }
  }
}